modules/javafx.controls/src/test/java/test/com/sun/javafx/scene/control/SelectedItemsReadOnlyObservableListTest.java

```diff
@@ -0,0 +1,221 @@
+    List<String> changes;
+
+    @Before
+    public void setup() {
+        changes = new ArrayList<>();
+        items = FXCollections.observableArrayList("foo", "bar", "baz", "qux", "quz");
+        selectedIndices = new TestObservableList<>();
+        selectedItems = new SelectedItemsReadOnlyObservableList<>(selectedIndices, () -> 0) {
+            @Override protected String getModelItem(int index) { return items.get(index); }
+        };
+        selectedItems.addListener((ListChangeListener<? super String>) c -> changes.add(c.toString()));
+    }
+
+    /**
+     * { [foo, bar, baz, qux, quz] added at 0 }
+     * { [foo, bar, baz, qux, quz] removed at 0 }
+     */
+    @Test
+    public void testAddAndRemoveEntireRange() {
+        selectedIndices.addAll(0, 1, 2, 3, 4);
+        assertEquals(1, changes.size());
+        assertEquals(change(added(0, "foo", "bar", "baz", "qux", "quz")), changes.get(0));
+        changes.clear();
+        selectedIndices.removeAll(0, 1, 2, 3, 4);
+        assertEquals(1, changes.size());
+        assertEquals(change(removed(0, "foo", "bar", "baz", "qux", "quz")), changes.get(0));
+    }
+
+    /**
+     * { [foo, bar, baz, qux, quz] added at 0 }
+     * { [bar, baz, qux] removed at 1 }
+     */
+    @Test
+    public void testRemoveInteriorRange() {
+        selectedIndices.addAll(0, 1, 2, 3, 4);
+        changes.clear();
+        selectedIndices.removeAll(1, 2, 3);
+        assertEquals(1, changes.size());
+        assertEquals(change(removed(1, "bar", "baz", "qux")), changes.get(0));
+    }
+
+    /**
+     * { [foo, bar, baz, qux, quz] added at 0 }
+     * { [foo, bar] removed at 0, [qux, quz] removed at 1 }
+     */
+    @Test
+    public void testRemoveDisjointRanges() {
+        selectedIndices.addAll(0, 1, 2, 3, 4);
+        changes.clear();
+        selectedIndices.removeAll(0, 1, 3, 4);
+        assertEquals(1, changes.size());
+        assertEquals(change(
+            removed(0, "foo", "bar"),
+            removed(1, "qux", "quz")), changes.get(0));
+    }
+
+    /**
+     * Note that using non-atomic swap operations on 'selectedIndices' doesn't work for
+     * SelectedItemsReadOnlyObservableList, since it results in intermediate states where
+     * 'selectedIndices' contains duplicates. This results in incorrect change notifications,
+     * which can be seen in this test:
+     *
+     * { [foo, bar] added at 0 }
+     * { [foo] replaced by [bar] at 0 }, but should be: [foo, bar] replaced by [bar, foo]
+     *
+     * This test is a documentation of that fact, and it will start to fail if replaceAll
+     * is implemented as an atomic operation.
+     */
+    @Test
+    public void testReplaceTwoItems() {
+        selectedIndices.addAll(0, 1);
+        changes.clear();
+        selectedIndices.replaceAll(i -> i == 0 ? 1 : 0);
+        assertEquals(1, changes.size());
+        assertEquals(change(replaced(0, range("foo"), range("bar"))), changes.get(0));
+    }
+
```

Comment:
Another thingy I don't quite understand, sry ;) 

What do you want to test here? We have:

selectedItems before: [foo, bar]
selectedItems after: [bar, foo]

so we need a change (the "should be" snippet in the description of this method)

replaced [foo, bar] by [bar, foo] at 0  

we get: 

replaced [foo] by [bar] at 0

which is an incorrect notification (because it's incomplete, doesn't notify at all about the replace at 1) covered by a passing test .. feels fishy. The other way round: if selectedItems cannot react to such a use-case with a correct notification, it might be broken?

Reply:
Yes, `selectedItems` is broken. I agree that having an objectively incorrect test pass is fishy, but the alternative would be to not have the test at all, and lose the "free" heads-up if at some point in the future the underlying issue is fixed. Do you think it's better to have the correct test, but ignore it?

Reply:
well, false greens are the worst we can have, IMO :). I think the test should concentrate on the isolated selectedItems behavior which __must__ propagate all notifications that it receives (from indices) in terms of items. If it fails to do so, it's a failure of its own implementation and should produce a failing test. 

Consider two scenarios: 

A: indices firing 2 separate changes

        selectedIndices.replaceAll(i -> i == 0 ? 1 : 0);
        assertEquals(2, changes.size());
        assertEquals(change(replaced(0, range("foo"), range("bar"))), changes.get(0));
        assertEquals(change(replaced(1, range("bar"), range("foo"))), changes.get(1));

B: indices firing a composed change:

        selectedIndices._beginChange();
        selectedIndices.replaceAll(i -> i == 0 ? 1 : 0);
        selectedIndices._endChange();
        assertEquals(indicesChanges.size(), changes.size());
        assertEquals(1, changes.size());
        assertEquals(change(replaced(0, range("foo", "bar"), range("bar", "foo"))), changes.get(0));

B passes both before and after the fix, A fails both before and after the fix. Whether or not that can be helped currently, is a different story. If can't be solved right now, I would suggest to keep it failing, file an issue about it and ignore the test with the issue id.

Reply:
I disabled the tests until the underlying [issue](https://bugs.openjdk.java.net/browse/JDK-8267951) is fixed.

Reply:
Would suggest to make them correct before ignoring - expected are the exact analogue of those fired by selectedIndices (which is two in the test setup)

Reply:
Done.